aix: protect uv_exepath() from uv_set_process_title()

richardlau opened the pull request:
Store a copy of the original argv[0] to protect `uv_exepath()`
against `uv_set_process_title()` changing the value of argv[0].

Since `uv_exepath()` is in `src/unix/aix-common.c`.
cc @dmabupt re. IBMi

Changes assume it is possible for `uv_setup_args()` not to 
be called prior to `uv_exepath()` being called (current 
[documentation for `uv_setup_args()`](http://docs.libuv.org/en/v1.x/misc.html#c.uv_setup_args) says that it is
"Required for getting / setting the process title.") in 
which case the existing code is used. Our test runner always
calls `uv_setup_args()` prior to running any tests.

Fixes: https://github.com/libuv/libuv/issues/2674
CI: https://ci.nodejs.org/job/libuv-test-commit/1985/

bnoordhuis commented on uv.gyp:
```diff
@@ -305,7 +305,6 @@
                 'src/unix/ibmi.c',
                 'src/unix/posix-poll.c',
                 'src/unix/no-fsevents.c',
-                'src/unix/no-proctitle.c',
```
Also needs an update to Makefile.am:
```diff
diff --git a/Makefile.am b/Makefile.am
index 9a06b9ce..f2bd4cee 100644
--- a/Makefile.am
+++ b/Makefile.am
@@ -379,8 +379,7 @@ uvinclude_HEADERS += include/uv/posix.h
 libuv_la_SOURCES += src/unix/aix-common.c \
                     src/unix/ibmi.c \
                     src/unix/posix-poll.c \
-                    src/unix/no-fsevents.c \
-                    src/unix/no-proctitle.c
+                    src/unix/no-fsevents.c
 endif
 
 if ANDROID
```

bnoordhuis commented on src/unix/aix-common.c:
```diff
@@ -119,7 +126,7 @@ int uv_exepath(char* buffer, size_t* size) {
     char *clonedpath = NULL;
     char *token = NULL;
-    char *path = getenv("PATH");
+    char *path = original_env_path != NULL ? original_env_path : getenv("PATH");
```
Long line. I'd write it like this:
```c
char* path;

path = original_env_path;
if (path == NULL)
  path = getenv("PATH");
```

bnoordhuis commented on test/test-get-currentexe.c:
```diff
@@ -82,5 +82,27 @@ TEST_IMPL(get_currentexe) {
+  /* Verify uv_exepath is not affected by uv_set_process_title(). */
+  r = uv_set_process_title("foobar");
+  ASSERT(r == 0);
+  size = sizeof(buffer) / sizeof(buffer[0]);
```
```suggestion
  size = sizeof(buffer);
```
`sizeof(char)` is always 1.

richardlau replied:
With the exception of the `uv_set_process_title()` call these lines were copied from earlier in the test. I’ll change those as well for consistency.

bnoordhuis commented on test/test-get-currentexe.c:
```diff
@@ -82,5 +82,27 @@ TEST_IMPL(get_currentexe) {
+  ASSERT(r == 0);
+  size = sizeof(buffer) / sizeof(buffer[0]);
+  r = uv_exepath(buffer, &size);
+  ASSERT(!r);
```
```suggestion
  ASSERT(r == 0);
```

bnoordhuis commented on test/test-get-currentexe.c:
```diff
@@ -82,5 +82,27 @@ TEST_IMPL(get_currentexe) {
+    path = executable_path + 2;
+  } else {
+    path = executable_path;
+  }
```
No braces around single-line consequents/alternates. Also, a slightly more succinct way of writing it:
```c
path = executable_path;
if (path[0] == '.' && path[1] == '/')
  path += 2;
```

bnoordhuis commented on test/test-get-currentexe.c:
```diff
@@ -82,5 +82,27 @@ TEST_IMPL(get_currentexe) {
+  /* Verify that the path returned from uv_exepath is a subdirectory of
+   * executable_path.
+   */
+  ASSERT(match && !strcmp(match, path));
```
```suggestion
  ASSERT(match != NULL && !strcmp(match, path));
```

dmabupt commented:
Hello @richardlau,
The code change works fine on IBMi PASE -- `ok 111 - get_currentexe`.

vtjnash commented:
IIUC, it is now getting stale, waiting for minor nits to be addressed?

richardlau commented:
Yes, this one is with me to address nits. This PR does address an issue on AIX but is not urgent as it doesn’t affect Node.js.

richardlau commented:
Sorry this has taken so long to get back to. I've rebased, updated to use the new assert macros in the changed test and believe I've addressed @bnoordhuis 's suggestions.

CI is green: https://ci.nodejs.org/job/libuv-test-commit/1985/

richardlau commented:
Landed in https://github.com/libuv/libuv/commit/ea92e9c720376470676582c3ac1bc9921c9142e7.
